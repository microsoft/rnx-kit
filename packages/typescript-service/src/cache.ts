import ts from "typescript";
import { VersionedSnapshot } from "./snapshot";

export class ProjectFileCache {
  private files: Map<string, VersionedSnapshot> = new Map();

  constructor(fileNames: string[]) {
    fileNames.forEach((fileName) => this.add(fileName));
  }

  add(fileName: string) {
    this.files.set(fileName, new VersionedSnapshot(fileName));
  }

  update(fileName: string, snapshot?: ts.IScriptSnapshot) {
    if (!this.files.has(fileName)) {
      throw new Error(`Cannot update unknown project file ${fileName}`);
    }
    this.files.get(fileName)!.update(snapshot);
  }

  delete(fileName: string) {
    this.files.delete(fileName);
  }

  getFileNames(): string[] {
    return Array.from(this.files.keys());
  }

  getVersion(fileName: string): string | undefined {
    return this.files.get(fileName)?.getVersion();
  }

  getSnapshot(fileName: string): ts.IScriptSnapshot | undefined {
    return this.files.get(fileName)?.getSnapshot();
  }
}

// TODO: the files in this list need to be watched. on change/delete, remove from this list and let the file be re-cached on demand
export class ExternalFileCache {
  private files: Map<string, VersionedSnapshot> = new Map();

  getSnapshot(fileName: string): ts.IScriptSnapshot {
    if (!this.files.has(fileName)) {
      this.files.set(fileName, new VersionedSnapshot(fileName));
    }

    return this.files.get(fileName)!.getSnapshot();
  }
}
