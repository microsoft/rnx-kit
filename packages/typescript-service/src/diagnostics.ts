import ts from "typescript";
import { getCanonicalFileName, getNewLine } from "./util";

export type DiagnosticWriter = {
  format: (diagnostic: ts.Diagnostic) => string;
  print: (diagnostic: ts.Diagnostic) => void;
};

export function createDiagnosticWriter(
  write?: (message: string) => void
): DiagnosticWriter {
  const writeDiagnostic = write ?? ts.sys.write;

  const formatDiagnosticsHost: ts.FormatDiagnosticsHost = {
    getCurrentDirectory: ts.sys.getCurrentDirectory,
    getCanonicalFileName,
    getNewLine,
  };

  function format(diagnostic: ts.Diagnostic): string {
    return ts.formatDiagnosticsWithColorAndContext(
      Array.isArray(diagnostic) ? diagnostic : [diagnostic],
      formatDiagnosticsHost
    );
  }

  function print(diagnostic: ts.Diagnostic) {
    const message = format(diagnostic);
    writeDiagnostic(message);
  }

  return {
    format,
    print,
  };
}
