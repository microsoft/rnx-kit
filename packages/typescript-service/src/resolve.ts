import ts from "typescript";
import { getCanonicalFileName } from "./util";

export type Resolvers = {
  resolveModuleNames: (
    moduleNames: string[],
    containingFile: string,
    _reusedNames: string[] | undefined,
    redirectedReference?: ts.ResolvedProjectReference
  ) => (ts.ResolvedModuleFull | undefined)[];

  getResolvedModuleWithFailedLookupLocationsFromCache: (
    modulename: string,
    containingFile: string
  ) => ts.ResolvedModuleWithFailedLookupLocations | undefined;

  resolveTypeReferenceDirectives: (
    typeDirectiveNames: string[],
    containingFile: string,
    redirectedReference?: ts.ResolvedProjectReference
  ) => (ts.ResolvedTypeReferenceDirective | undefined)[];
};

export function createResolvers(options: ts.CompilerOptions): Resolvers {
  const moduleResolutionCache = ts.createModuleResolutionCache(
    ts.sys.getCurrentDirectory(),
    getCanonicalFileName,
    options
  );

  const moduleResolutionHost: ts.ModuleResolutionHost = {
    fileExists: ts.sys.fileExists,
    readFile: ts.sys.readFile,
  };

  function resolveModuleNames(
    moduleNames: string[],
    containingFile: string,
    _reusedNames: string[] | undefined,
    redirectedReference?: ts.ResolvedProjectReference
  ): (ts.ResolvedModuleFull | undefined)[] {
    const resolvedModules: (ts.ResolvedModuleFull | undefined)[] = [];
    for (const moduleName of moduleNames) {
      // try to use standard resolution
      const result = ts.resolveModuleName(
        moduleName,
        containingFile,
        options,
        moduleResolutionHost,
        moduleResolutionCache,
        redirectedReference
      );
      resolvedModules.push(result.resolvedModule);
    }
    return resolvedModules;
  }

  // TODO: implement this
  function getResolvedModuleWithFailedLookupLocationsFromCache(
    _modulename: string,
    _containingFile: string
  ): ts.ResolvedModuleWithFailedLookupLocations | undefined {
    throw new Error("Not implemented");
  }

  function resolveTypeReferenceDirectives(
    typeDirectiveNames: string[],
    containingFile: string,
    redirectedReference?: ts.ResolvedProjectReference
  ): (ts.ResolvedTypeReferenceDirective | undefined)[] {
    const resolved: (ts.ResolvedTypeReferenceDirective | undefined)[] = [];
    for (const name of typeDirectiveNames) {
      const result = ts.resolveTypeReferenceDirective(
        name,
        containingFile,
        options,
        moduleResolutionHost,
        redirectedReference
      );
      resolved.push(result.resolvedTypeReferenceDirective);
    }
    return resolved;
  }

  return {
    resolveModuleNames,
    getResolvedModuleWithFailedLookupLocationsFromCache,
    resolveTypeReferenceDirectives,
  };
}
