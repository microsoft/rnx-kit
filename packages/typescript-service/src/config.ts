import * as ts from "typescript";
import { DiagnosticWriter } from "./diagnostics";
import { isNonEmptyArray } from "./util";

export type ProjectConfig = ts.ParsedCommandLine;

export class ProjectConfigLoader {
  private diagnosticWriter: DiagnosticWriter;
  private parseConfigFileHost: ts.ParseConfigFileHost;
  private extendedConfigCache: ts.ESMap<string, ts.ExtendedConfigCacheEntry>;

  constructor(diagnosticWriter: DiagnosticWriter) {
    this.diagnosticWriter = diagnosticWriter;
    this.parseConfigFileHost = {
      useCaseSensitiveFileNames: ts.sys.useCaseSensitiveFileNames,
      readDirectory: ts.sys.readDirectory,
      fileExists: ts.sys.fileExists,
      readFile: ts.sys.readFile,
      getCurrentDirectory: ts.sys.getCurrentDirectory,
      onUnRecoverableConfigFileDiagnostic: diagnosticWriter.print,
    };
    this.extendedConfigCache = new Map();
  }

  public load(
    searchPath: string,
    fileName: string = "tsconfig.json"
  ): ProjectConfig {
    const configFileName = ts.findConfigFile(
      searchPath,
      ts.sys.fileExists,
      fileName
    );
    if (!configFileName) {
      throw new Error(
        `Failed to find '${fileName}' under search-path '${searchPath}'`
      );
    }

    const commandLine = ts.getParsedCommandLineOfConfigFile(
      configFileName,
      {}, // optionsToExtend
      this.parseConfigFileHost,
      this.extendedConfigCache
    );
    if (!commandLine) {
      throw new Error(`Failed to load '${fileName}' (${configFileName})`);
    }

    if (isNonEmptyArray(commandLine.errors)) {
      this.diagnosticWriter.print(commandLine.errors);
      throw new Error(`Failed to load '${fileName}' (${configFileName})`);
    }

    return commandLine;
  }
}
