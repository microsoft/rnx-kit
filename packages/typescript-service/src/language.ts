import * as fs from "fs";
import * as ts from "typescript";
import { IResolvers } from "./resolve";

export function createLanguageService(
  documentRegistry: ts.DocumentRegistry,
  options: ts.CompilerOptions,
  resolvers: IResolvers
) {
  const fileVersions: ts.MapLike<number> = {};

  const languageServiceHost: ts.LanguageServiceHost = {
    getScriptFileNames: () => Object.keys(fileVersions),
    getScriptVersion: (fileName) =>
      fileVersions.hasOwnProperty(fileName)
        ? fileVersions[fileName].toString()
        : "0",
    getScriptSnapshot: (fileName) => {
      if (!fs.existsSync(fileName)) {
        return undefined;
      }
      return ts.ScriptSnapshot.fromString(fs.readFileSync(fileName).toString());
    },
    getCurrentDirectory: () => process.cwd(),
    getCompilationSettings: () => options,
    getDefaultLibFileName: (o) => ts.getDefaultLibFilePath(o),
    fileExists: ts.sys.fileExists,
    readFile: ts.sys.readFile,
    readDirectory: ts.sys.readDirectory,
    directoryExists: ts.sys.directoryExists,
    getDirectories: ts.sys.getDirectories,

    resolveModuleNames: resolvers.resolveModuleNames,
    resolveTypeReferenceDirectives: resolvers.resolveTypeReferenceDirectives,
  };

  const languageService = ts.createLanguageService(
    languageServiceHost,
    documentRegistry
  );

  function validateFile(fileName: string) {
    return languageService
      .getCompilerOptionsDiagnostics()
      .concat(languageService.getSyntacticDiagnostics(fileName))
      .concat(languageService.getSemanticDiagnostics(fileName));
  }

  function loadFile(fileName: string, snapshot?: ts.IScriptSnapshot) {
    const exists = fileVersions.hasOwnProperty(fileName);
    const finalSnapshot =
      snapshot ||
      ts.ScriptSnapshot.fromString(fs.readFileSync(fileName).toString());

    if (!exists) {
      fileVersions[fileName] = 1;
      documentRegistry.acquireDocument(
        fileName,
        options,
        finalSnapshot,
        fileVersions[fileName].toString()
      );
    } else {
      fileVersions[fileName]++;
      documentRegistry.updateDocument(
        fileName,
        options,
        finalSnapshot,
        fileVersions[fileName].toString()
      );
    }
  }

  return {
    validateFile,
    loadFile,
  };
}
