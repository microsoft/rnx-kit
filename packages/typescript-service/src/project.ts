import * as ts from "typescript";
import { ProjectFileCache, ExternalFileCache } from "./cache";
import { Resolvers } from "./resolve";
import { ProjectConfig } from "./config";

export class Project {
  private resolvers: Resolvers;
  private projectConfig: ProjectConfig;

  private projectFiles: ProjectFileCache;
  private externalFiles: ExternalFileCache;

  private languageService: ts.LanguageService;

  constructor(
    documentRegistry: ts.DocumentRegistry,
    resolvers: Resolvers,
    projectConfig: ProjectConfig
  ) {
    this.resolvers = resolvers;
    this.projectConfig = projectConfig;

    this.projectFiles = new ProjectFileCache(projectConfig.fileNames);
    this.externalFiles = new ExternalFileCache();

    const languageServiceHost: ts.LanguageServiceHost = {
      getCompilationSettings: () => this.projectConfig.options,
      //getNewLine?(): string;
      //getProjectVersion?(): string;
      getScriptFileNames: () => this.projectFiles.getFileNames(),
      //getScriptKind?(fileName: string): ts.ScriptKind;
      getScriptVersion: (fileName) =>
        this.projectFiles.getVersion(fileName) ?? "0",
      getScriptSnapshot: (fileName) =>
        this.projectFiles.getSnapshot(fileName) ??
        this.externalFiles.getSnapshot(fileName),
      //getProjectReferences?(): readonly ProjectReference[] | undefined;
      //getLocalizedDiagnosticMessages?(): any;
      //getCancellationToken?(): HostCancellationToken;
      getCurrentDirectory: () => process.cwd(),
      getDefaultLibFileName: (o) => ts.getDefaultLibFilePath(o),
      //log?(s: string): void;
      //trace?(s: string): void;
      //error?(s: string): void;
      //useCaseSensitiveFileNames?(): boolean;

      /*
       * LS host can optionally implement these methods to support completions for module specifiers.
       * Without these methods, only completions for ambient modules will be provided.
       */
      readDirectory: ts.sys.readDirectory,
      readFile: ts.sys.readFile,
      realpath: ts.sys.realpath,
      fileExists: ts.sys.fileExists,

      /*
       * LS host can optionally implement these methods to support automatic updating when new type libraries are installed
       */
      //getTypeRootsVersion?(): number;

      /*
       * LS host can optionally implement this method if it wants to be completely in charge of module name resolution.
       * if implementation is omitted then language service will use built-in module resolution logic and get answers to
       * host specific questions using 'getScriptSnapshot'.
       *
       * If this is implemented, `getResolvedModuleWithFailedLookupLocationsFromCache` should be too.
       */
      resolveModuleNames: this.resolvers.resolveModuleNames,
      getResolvedModuleWithFailedLookupLocationsFromCache:
        this.resolvers.getResolvedModuleWithFailedLookupLocationsFromCache,
      resolveTypeReferenceDirectives:
        this.resolvers.resolveTypeReferenceDirectives,

      /*
       * Required for full import and type reference completions.
       * These should be unprefixed names. E.g. `getDirectories("/foo/bar")` should return `["a", "b"]`, not `["/foo/bar/a", "/foo/bar/b"]`.
       */
      getDirectories: ts.sys.getDirectories,

      /**
       * Gets a set of custom transformers to use during emit.
       */
      //getCustomTransformers?(): CustomTransformers | undefined;

      //isKnownTypesPackageName?(name: string): boolean;
      //installPackage?(options: InstallPackageOptions): Promise<ApplyCodeActionCommandResult>;
      //writeFile?(fileName: string, content: string): void;

      //getParsedCommandLine?(fileName: string): ParsedCommandLine | undefined;

      directoryExists: ts.sys.directoryExists,
    };

    this.languageService = ts.createLanguageService(
      languageServiceHost,
      documentRegistry
    );
  }

  warmup() {
    this.languageService.getProgram();
  }

  validateFile(fileName: string): ts.Diagnostic[] {
    const syntax = this.languageService.getSyntacticDiagnostics(
      fileName
    ) as ts.Diagnostic[];
    const semantics = this.languageService.getSemanticDiagnostics(fileName);
    return syntax.concat(semantics);
  }

  validate(): ts.Diagnostic[] {
    const diagnostics: ts.Diagnostic[] = [];
    for (const fileName of this.projectFiles.getFileNames()) {
      Array.prototype.push.apply(diagnostics, this.validateFile(fileName));
    }
    return diagnostics;
  }

  addFile(fileName: string) {
    this.projectFiles.add(fileName);
  }

  updateFile(fileName: string, snapshot?: ts.IScriptSnapshot) {
    this.projectFiles.update(fileName, snapshot);
  }

  removeFile(fileName: string) {
    this.projectFiles.delete(fileName);
  }
}
