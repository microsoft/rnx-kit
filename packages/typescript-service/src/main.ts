import * as ts from "typescript";
import { ProjectConfigLoader } from "./config";
import { Project } from "./project";
import { createResolvers } from "./resolve";
import { createDiagnosticWriter } from "./diagnostics";
import { isNonEmptyArray } from "./util";

// TODO: add documentation (README.md)
// TODO: remove this file and hook it from index.ts
export function main() {
  const diagnosticWriter = createDiagnosticWriter();

  const projectConfigLoader = new ProjectConfigLoader(diagnosticWriter);
  const projectConfig = projectConfigLoader.load("./");

  const resolvers = createResolvers(projectConfig.options);

  const documentRegistry = ts.createDocumentRegistry();

  const project = new Project(documentRegistry, resolvers, projectConfig);
  project.warmup();

  const f = projectConfig.fileNames[0];

  // typecheck this file -- will succeed
  console.log("validating file: %o", f);
  let diagnostics = project.validateFile(f);
  if (diagnostics.length > 0) {
    diagnosticWriter.print(diagnostics);
  }

  // replace this file with some questionable code to generate errors
  console.log("re-loading file: %o", f);
  project.updateFile(
    f,
    ts.ScriptSnapshot.fromString("function foo(x: any) { return 1; }")
  );

  // typecheck again -- will fail
  console.log("validating file: %o", f);
  diagnostics = project.validateFile(f);
  if (isNonEmptyArray(diagnostics)) {
    diagnosticWriter.print(diagnostics);
  }

  console.log("removing file: %o", f);
  project.removeFile(f);

  console.log("re-adding file: %o", f);
  project.addFile(f);

  console.log("re-validating file: %o", f);
  diagnostics = project.validateFile(f);
  if (isNonEmptyArray(diagnostics)) {
    diagnosticWriter.print(diagnostics);
  }
}
