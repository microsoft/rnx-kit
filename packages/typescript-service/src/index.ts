export * from "./cache";
export * from "./config";
export * from "./diagnostics";
export * from "./project";
export * from "./resolve";
export * from "./snapshot";

import { main } from "./main";
main();
