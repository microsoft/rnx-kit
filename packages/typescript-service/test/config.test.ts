import path from "path";
import ts from "typescript";
import { DiagnosticWriter } from "../src/diagnostics";
import { ProjectConfigLoader } from "../src/config";

describe("ProjectConfigLoader", () => {
  const writer: DiagnosticWriter = {
    format: jest.fn(),
    print: jest.fn(),
  };

  const fixturePath = path.join(process.cwd(), "test", "__fixtures__");

  afterEach(() => {
    jest.resetAllMocks();
  });

  test("load throws when it can't find a config file", () => {
    const loader = new ProjectConfigLoader(writer);
    expect(() =>
      loader.load(fixturePath, "invalid-config-file-name")
    ).toThrowErrorMatchingSnapshot();
  });

  test("load throws when it encounters an invalid config file", () => {
    const loader = new ProjectConfigLoader(writer);
    expect(() =>
      loader.load(fixturePath, "invalid-tsconfig.json")
    ).toThrowErrorMatchingSnapshot();
  });

  test("load prints at least one diagnostic when it encounters an invalid config file", () => {
    const loader = new ProjectConfigLoader(writer);
    try {
      loader.load(fixturePath, "invalid-tsconfig.json");
    } catch {}
    // @ts-ignore
    expect(writer.print.mock.calls.length).toBeGreaterThan(0);
  });

  test("load returns a valid config", () => {
    const loader = new ProjectConfigLoader(writer);
    const config = loader.load(fixturePath, "valid-tsconfig.json");
    expect(config.options.target).toEqual(ts.ScriptTarget.ES2015);
    expect(config.options.module).toEqual(ts.ModuleKind.CommonJS);
  });
});
