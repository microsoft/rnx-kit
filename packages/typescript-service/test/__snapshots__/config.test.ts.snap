// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectConfigLoader load throws when it can't find a config file 1`] = `"Failed to find 'invalid-config-file-name' under search-path '/Users/afoxman/repos/rnx-kit/packages/typescript-service/test/__fixtures__'"`;

exports[`ProjectConfigLoader load throws when it encounters an invalid config file 1`] = `"Failed to load 'invalid-tsconfig.json' (/Users/afoxman/repos/rnx-kit/packages/typescript-service/test/__fixtures__/invalid-tsconfig.json)"`;
